### PR TITLE
Initial support for loongarch64

### PR DESCRIPTION
**Pull request by qiangxuhui**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Experimentally support loongarch64 as a new platform for ClickHouse.

### Documentation entry for user-facing changes

- [X] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

<details>

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 537b56b1396d19ab3caa0bf546a2c546ae923dbf with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63733/537b56b1396d19ab3caa0bf546a2c546ae923dbf/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -247,6 +247,14 @@ jobs:
       build_name: binary_s390x
       data: ${{ needs.RunConfig.outputs.data }}
       checkout_depth: 0
+  BuilderBinLOONGARCH64:
```
Comment (rschu1ze) [inline .github/workflows/master.yml:250]: Cosmetic: Could we rename that to `BuilderBinLoongarch64`?

File: docker/packager/packager [hunk 2]
```diff
@@ -149,6 +149,7 @@ def parse_env_variables(
     PPC_SUFFIX = "-ppc64le"
     RISCV_SUFFIX = "-riscv64"
     S390X_SUFFIX = "-s390x"
+    LOONGARCH_SUFFIX = "-loongarch64"
```
Comment (rschu1ze) [inline docker/packager/packager:152]: It looks as if more changes are required in this file to enable CI builds for LoogArch64. Please see https://github.com/ClickHouse/ClickHouse/pull/63135/files#diff-00c66281b1b2973ae73eab7b260f1ef5ecc0a72ce4cec1b8fa71922d31c1f9dd for a related PR (about s390x).

File: src/Common/ThreadFuzzer.cpp [hunk 3]
```diff
@@ -361,7 +361,7 @@ void ThreadFuzzer::setup() const
 
 /// Starting from glibc 2.34 there are no internal symbols without version,
 /// so not __pthread_mutex_lock but __pthread_mutex_lock@2.2.5
-#if defined(OS_LINUX) and !defined(USE_MUSL)
+#if defined(OS_LINUX) and !defined(USE_MUSL) and !defined(__loongarch64)
```
Comment (rschu1ze) [inline src/Common/ThreadFuzzer.cpp:364]: It is not really clear to me why the special handling for loongarch64 is needed here (l. 364, l. 387). Can't we define a normal branch instead (like l. 374)?

File: cmake/target.cmake [hunk 4]
```diff
@@ -61,6 +61,21 @@ if (CMAKE_CROSSCOMPILING)
...
+            set (ENABLE_PARQUET OFF CACHE INTERNAL "")
+            set (ENABLE_GRPC OFF CACHE INTERNAL "")
+            set (ENABLE_HDFS OFF CACHE INTERNAL "")
+            set (ENABLE_MYSQL OFF CACHE INTERNAL "")
+            set (ENABLE_RUST OFF CACHE INTERNAL "")
+            set (ENABLE_LIBPQXX OFF CACHE INTERNAL "")
+            set (ENABLE_EMBEDDED_COMPILER OFF CACHE INTERNAL "")
+            set (ENABLE_DWARF_PARSER OFF CACHE INTERNAL "")
+            set (ENABLE_BLAKE3 OFF CACHE INTERNAL "")
+            set (COMPILER_CACHE disabled CACHE INTERNAL "")
```
Comment (rschu1ze) [inline cmake/target.cmake:78]: Disabling `COMPILER_CACHE` may negatively affect compilation times for LoongArch64 builds in CI. Could we remove l. 78?

File: src/Common/ThreadFuzzer.cpp (hunk 3 above)
Comment (qiangxuhui) [reply src/Common/ThreadFuzzer.cpp:364]: In glibc>=2.36, '__pthread_mutex_lock' and '__pthread_mutex_unlock' do not exist:

```sh
# s390's pthread_mutex_lock symbol table
# readelf -sW contrib/sysroot/linux-s390x/s390x-linux-gnu/libc/lib64/libpthread.so.0 | grep  "pthread_mutex_lock"
   168: 000000000000a0c0   760 FUNC    WEAK   DEFAULT   13 pthread_mutex_lock@@GLIBC_2.2
   199: 000000000000a0c0   760 FUNC    GLOBAL DEFAULT   13 __pthread_mutex_lock@@GLIBC_2.2
    39: 0000000000009b58  1380 FUNC    LOCAL  DEFAULT   13 __pthread_mutex_lock_full
   195: 000000000000a0c0   760 FUNC    LOCAL  DEFAULT   13 __GI___pthread_mutex_lock
   436: 000000000000a0c0   760 FUNC    WEAK   DEFAULT   13 pthread_mutex_lock
   563: 000000000000a0c0   760 FUNC    GLOBAL DEFAULT   13 __pthread_mutex_lock

# loongarch64's pthread_mutex_lock symbol table
# readelf -sW contrib/sysroot/linux-loongarch64/lib/loongarch64-linux-gnu/libc.so.6 | grep  "pthread_mutex_lock"
  1265: 0000000000081578   828 FUNC    GLOBAL DEFAULT   13 pthread_mutex_lock@@GLIBC_2.36
```

The minimum version of glibc for 'loongarch64' is 2.36, so we can't define a normal branch instead (like l. 374).

File: .github/workflows/master.yml (hunk 1 above)
Comment (qiangxuhui) [reply .github/workflows/master.yml]: Yes.

File: cmake/target.cmake (hunk 4 above)
Comment (qiangxuhui) [reply cmake/target.cmake]: I have removed it.

**Comment by alexey-milovidov** [comment]
My request - when writing not in lowercase, write it as `LoongArch64`, not `LOONGARCH64`

**Comment by qiangxuhui** [comment]
@alexey-milovidov I replaced `LOONGARCH64` with `Loongarch64` in some files. The reason why the remaining files which contain `LOONGARCH64` but have not been modified is to be consistent with other platforms.

Please take a look.

**Comment by alexey-milovidov** [comment]
The patch for libunwind is merged; you can update the submodule to master.

**Comment by qiangxuhui** [comment]
I have updated the submodule `contrib/libunwind` to master, but there are still code conflicts, Do I need to rebase master to resolve the code conflicts?

**Comment by qiangxuhui** [comment]
I have updated the submodule `contrib/libunwind` to master, but there are still code conflicts, Do I need to rebase master to resolve the code conflicts?

**Comment by qiangxuhui** [comment]
> I just merged from `master`, that resolves the issue.
> 
> @qiangxuhui Let's also avoid force-pushing once the code has been reviewed. A force-push allows to rewrite the entire commit history, meaning that all previously reviewed code (20k LOC in this PR) need to be re-reviewed.

@rschu1ze Thank you, I will follow your suggestions in future code push. What should I do next?

**Comment by rschu1ze** [comment]
Let's wait for results of CI builds/tests, if green we can merge.

**Comment by qiangxuhui** [comment]
@rschu1ze `AST fuzzer (debug)` and `Stress test (debug)` failed, do I need to fix them?

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**